src/sentry/sentry_metrics/consumers/indexer/routing_producer.py

```diff
@@ -0,0 +1,166 @@
+    routed to.
+    """
+
+    producer: Producer
+    topic: Topic
+
+
+class MessageRouter(ABC):
+    """
+    An abstract class that defines the interface for a message router. A message
+    router relies on the implementation of the get_route_for_message method
+    to perform the actual routing.
+
+    Some examples of what message routers could do include:
+    1. A message router that routes messages to a single topic. This is
+    similar to not having a message router at all.
+    2. A message router that performs round robin routing to a set of topics.
+    3. A message router that routes messages to a topic based on which slice
+    the message needs to be sent to.
+    """
+
+    @abstractmethod
+    def get_route_for_message(self, message: Message[RoutingPayload]) -> MessageRoute:
+        """
+        This method must return the MessageRoute on which the message should
+        be produced. Implementations of this method can vary based on the
+        specific use case.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def shutdown(self) -> None:
+        """
+        Shutdown the router. This method can be used to perform any cleanup
+        needed by the router.
+        """
+        raise NotImplementedError
+
+
+class RoutingProducerStep(ProcessingStrategy[RoutingPayload]):
+    """
+    This strategy is used to route messages to different producers/topics
+    based on the message payload. It delegates the routing logic to the
+    message router class which is passed in as a parameter. Please look at
+    the documentation for the MessageRouter class for more details.
+
+    The strategy also does not own any producers. The producers are owned by
+    the message router.
+    """
+
+    def __init__(
+        self,
+        commit_function: Commit,
+        message_router: MessageRouter,
+    ) -> None:
+        self.__commit_function = commit_function
+        self.__message_router = message_router
+        self.__closed = False
+        self.__offsets_to_be_committed: MutableMapping[Partition, Position] = {}
+        self.__queue: Deque[Tuple[Partition, Position, Future[Message[KafkaPayload]]]] = deque()
+
+    def poll(self) -> None:
+        """
+        Periodically check which messages have been produced successfully and
+        call the commit function for the offsets that have been processed.
+        In order to commit offsets we need strict ordering, hence we stop
+        processing the queue of pending futures as soon as we encounter one
+        which is not yet completed.
+        """
+        while self.__queue:
+            partition, position, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            future.result()
+
+            self.__queue.popleft()
+            self.__offsets_to_be_committed[partition] = position
+
+        self.__commit_function(self.__offsets_to_be_committed)
```

Comment:
```suggestion
        self.__commit_function({partition: position})
```
and remove `self.__offsets_to_be_committed` entirely

Reply:
Done.